File: ios/MenCare/Features/Admin/Manual/ManualBookingScreen.swift
import SwiftUI
import Observation

@MainActor
@Observable
final class ManualBookingViewModel {

    private let booking: BookingRepository
    private let crm: CrmRepository
    private let catalog: CatalogRepository

    var query = ""
    var results: [ClientRecord] = []
    var selectedClient: ClientRecord?
    var creatingClient = false
    var newFirst = ""
    var newLast = ""
    var newPhone = ""
    var newClientErrors: [String: String] = [:]
    var newClientError: String?
    var creatingClientInFlight = false
    var selectedOperatorId: String?
    var date: LocalDate = .today()
    var selectedServiceIds: [String] = []
    var slots: [LocalTime] = []
    var selectedSlot: LocalTime?
    /// Time the sheet was opened on (a tap on the agenda) or last picked:
    /// selected again as soon as the chosen services fit there.
    var preferredSlot: LocalTime?
    var sendSms = true
    var done = false

    let state = LoadState()
    private(set) var operators: [Operator] = []
    private(set) var services: [Service] = []
    var saveError: String?

    @ObservationIgnored private var searchTask: Task<Void, Never>?
    /// Operatore chiesto da chi apre la tendina (un tap sulla colonna): si
    /// applica appena la squadra e' arrivata.
    @ObservationIgnored private let requestedOperatorId: String?
    /// "Modifica" dall'agenda: l'appuntamento che questo modulo sostituisce.
    @ObservationIgnored private let editingId: String?

    var isEditing: Bool { editingId != nil }

    init(
        booking: BookingRepository, catalog: CatalogRepository, crm: CrmRepository,
        operatorId: String? = nil, date: LocalDate = .today(), time: LocalTime? = nil,
        editAppointment: Appointment? = nil, editClient: ClientRecord? = nil
    ) {
        self.booking = booking
        self.catalog = catalog
        self.crm = crm
        if let editAppointment {
            editingId = editAppointment.id
            requestedOperatorId = editAppointment.operatorId
            selectedOperatorId = editAppointment.operatorId
            self.date = editAppointment.date
            selectedServiceIds = editAppointment.serviceIds
            preferredSlot = editAppointment.time
            selectedClient = editClient
            query = editClient?.fullName ?? ""
        } else {
            editingId = nil
            requestedOperatorId = operatorId
            selectedOperatorId = operatorId
            self.date = date
            preferredSlot = time
        }
    }

    func load() async {
        await state.run {
            let snapshot = try await catalog.catalog()
            operators = snapshot.operators
            services = snapshot.services
            selectedOperatorId = requestedOperatorId ?? snapshot.operators.first?.id
        }
        refreshSlots()
    }

    var totalMinutes: Int {
        services.filter { selectedServiceIds.contains($0.id) }.reduce(0) { $0 + $1.durationMinutes }
    }

    var canSave: Bool {
        selectedClient != nil && selectedOperatorId != nil && !selectedServiceIds.isEmpty && selectedSlot != nil
    }

    /// The preferred time is taken for the chosen services.
    var preferredUnavailable: Bool {
        guard let preferredSlot, !selectedServiceIds.isEmpty else { return false }
        return selectedSlot == nil && !slots.contains(preferredSlot)
    }

    private var selectedOperator: Operator? {
        operators.first { $0.id == selectedOperatorId }
    }

    /// The chosen operator does not perform every selected service.
    var operatorIneligible: Bool {
        guard let op = selectedOperator else { return false }
        return selectedServiceIds.contains { !op.serviceIds.contains($0) }
    }

    /// The chosen operator has no working hours on the chosen day.
    var operatorOffDuty: Bool {
        guard let op = selectedOperator else { return false }
        return (op.weeklyHours[date.dayOfWeek] ?? []).isEmpty
    }

    func search(_ text: String) {
        query = text
        searchTask?.cancel()
        guard !text.trimmingCharacters(in: .whitespaces).isEmpty else {
            results = []
            return
        }
        // Ricerca sul server, con la pausa fra un tasto e l'altro.
        searchTask = Task {
            try? await Task.sleep(for: .milliseconds(300))
            guard !Task.isCancelled else { return }
            results = Array(((try? await crm.clients(query: text, segment: .tutti)) ?? []).prefix(5))
        }
    }

    func selectClient(_ client: ClientRecord) {
        selectedClient = client
        results = []
        query = client.fullName
        creatingClient = false
    }

    func startCreateClient() {
        creatingClient = true
        selectedClient = nil
    }

    func newClientFieldChanged() {
        newClientError = nil
        if !newClientErrors.isEmpty {
            newClientErrors = validateNewClient(firstName: newFirst, lastName: newLast, phone: newPhone)
        }
    }

    func createClient() {
        guard !creatingClientInFlight else { return }
        let errors = validateNewClient(firstName: newFirst, lastName: newLast, phone: newPhone)
        if !errors.isEmpty {
            // Prima il fallimento veniva ingoiato e il tasto sembrava rotto.
            newClientErrors = errors
            newClientError = L("validation_form_invalid")
            return
        }
        newClientErrors = [:]
        newClientError = nil
        creatingClientInFlight = true
        let phoneToSave = normalizePhone(newPhone) ?? newPhone
        Task {
            let result = await crm.createClient(
                firstName: newFirst.trimmingCharacters(in: .whitespaces),
                lastName: newLast.trimmingCharacters(in: .whitespaces),
                phone: phoneToSave
            )
            creatingClientInFlight = false
            switch result {
            case .success(let record):
                selectClient(record)
            case .failure:
                newClientError = L("validation_save_failed")
            }
        }
    }

    func selectOperator(_ operatorId: String) {
        selectedOperatorId = operatorId
        selectedSlot = nil
        refreshSlots()
    }

    func selectDate(_ newDate: LocalDate) {
        date = newDate
        selectedSlot = nil
        refreshSlots()
    }

    func toggleService(_ serviceId: String) {
        if let index = selectedServiceIds.firstIndex(of: serviceId) {
            selectedServiceIds.remove(at: index)
        } else {
            selectedServiceIds.append(serviceId)
        }
        selectedSlot = nil
        refreshSlots()
    }

    func selectSlot(_ slot: LocalTime) {
        selectedSlot = slot
        preferredSlot = slot
    }

    func refreshSlots() {
        Task {
            guard let selectedOperatorId, !selectedServiceIds.isEmpty else {
                slots = []
                return
            }
            // In modifica l'appuntamento non occupa il proprio posto: il suo orario resta sceglibile.
            slots = (try? await booking.availability(
                operatorId: selectedOperatorId, serviceIds: selectedServiceIds, date: date,
                ignoreAppointmentId: editingId
            ).slots) ?? []
            if selectedSlot == nil, let preferredSlot, slots.contains(preferredSlot) {
                selectedSlot = preferredSlot
            }
        }
    }

    func save() {
        guard let client = selectedClient, let slot = selectedSlot else { return }
        Task {
            saveError = nil
            // In modifica il server sostituisce il vecchio con il nuovo in un colpo
            // solo: se il nuovo non entra, il vecchio resta com'era.
            let result = await booking.book(
                BookingRequest(
                    clientId: client.id, operatorId: selectedOperatorId,
                    serviceIds: selectedServiceIds, start: date.atTime(slot), channel: .phone,
                    replacesAppointmentId: editingId
                )
            )
            switch result {
            case .success:
                done = true
            case .failure(let error):
                saveError = error.displayMessage
                refreshSlots()
            }
        }
    }
}

struct ManualBookingScreen: View {
    @Bindable var viewModel: ManualBookingViewModel
    let onDismiss: () -> Void

    @State private var showDatePicker = false
    @State private var headerHeight: CGFloat = 0
    @State private var formHeight: CGFloat = 0
    @State private var footerHeight: CGFloat = 0

    var body: some View {
        VStack(spacing: 0) {
            HStack {
                BarAction(text: L("manual_cancel"), action: onDismiss, color: .ink)
                    .frame(width: 72, alignment: .leading)
                Text(L(viewModel.isEditing ? "manual_title_edit" : "manual_title_short"))
                    .font(Typo.cormorant(21, weight: .regular))
                    .foregroundStyle(Color.ink)
                    .lineLimit(1)
                    .frame(maxWidth: .infinity)
                BarAction(text: L("manual_save"), action: viewModel.save, color: .oliveWood)
                    .frame(width: 72, alignment: .trailing)
            }
            .padding(.horizontal, 20)
            .padding(.vertical, 6)
            .measureHeight($headerHeight)

            ScrollView {
                VStack(alignment: .leading, spacing: 12) {
                    BrandSectionLabel(text: L("manual_client"))
                    // In modifica il cliente è quello dell'appuntamento: si mostra e basta.
                    if viewModel.isEditing {
                        FixedClientRow(client: viewModel.selectedClient)
                    } else {
                        FilledTextField(
                            label: "",
                            text: Binding(get: { viewModel.query }, set: { viewModel.search($0) }),
                            placeholder: L("manual_search_hint"),
                            leadingSystemImage: "magnifyingglass"
                        )
                        clientList
                    }

                    if viewModel.creatingClient {
                        NewClientForm(
                            firstLabel: L("manual_new_client_first"),
                            lastLabel: L("manual_new_client_last"),
                            phoneLabel: L("manual_new_client_phone"),
                            ctaLabel: L("manual_new_client"),
                            firstName: $viewModel.newFirst,
                            lastName: $viewModel.newLast,
                            phone: $viewModel.newPhone,
                            errors: viewModel.newClientErrors,
                            formError: viewModel.newClientError,
                            creating: viewModel.creatingClientInFlight,
                            onCreate: viewModel.createClient
                        )
                        .onChange(of: viewModel.newFirst) { viewModel.newClientFieldChanged() }
                        .onChange(of: viewModel.newLast) { viewModel.newClientFieldChanged() }
                        .onChange(of: viewModel.newPhone) { viewModel.newClientFieldChanged() }
                    }

                    HStack(alignment: .top, spacing: 10) {
                        VStack(alignment: .leading, spacing: 8) {
                            BrandSectionLabel(text: L("manual_operator"))
                            // Menu di sistema: tutti gli operatori, spunta su quello scelto.
                            Menu {
                                ForEach(viewModel.operators) { op in
                                    Button {
                                        viewModel.selectOperator(op.id)
                                    } label: {
                                        if op.id == viewModel.selectedOperatorId {
                                            Label(op.name, systemImage: "checkmark")
                                        } else {
                                            Text(op.name)
                                        }
                                    }
                                }
                            } label: {
                                dropdownLabel(
                                    viewModel.operators.first { $0.id == viewModel.selectedOperatorId }
                                        .map { shortOperatorName($0.name) } ?? "—"
                                )
                            }
                            .buttonStyle(.plain)
                        }
                        VStack(alignment: .leading, spacing: 8) {
                            BrandSectionLabel(text: L("manual_date"))
                            dropdownField(formatDateShort(viewModel.date)) {
                                showDatePicker = true
                            }
                        }
                    }

                    BrandSectionLabel(text: L("manual_services"))
                    // Tendina multi-selezione, gemella di quella dell'operatore.
                    MultiSelectDropdown(
                        options: viewModel.services.map { DropdownOption(id: $0.id, label: $0.name) },
                        selectedIds: viewModel.selectedServiceIds,
                        onToggle: viewModel.toggleService,
                        placeholder: L("manual_services_placeholder")
                    )

                    BrandSectionLabel(text: L("manual_slots", formatDuration(viewModel.totalMinutes)))
                    // Colonna vuota non vuol dire libera: se l'operatore non è in
                    // turno o non esegue i servizi scelti, lo si dice.
                    let emptyReason: String? = {
                        guard viewModel.selectedOperatorId != nil, !viewModel.selectedServiceIds.isEmpty else { return nil }
                        if viewModel.operatorIneligible { return L("manual_slots_ineligible") }
                        if viewModel.operatorOffDuty { return L("manual_slots_off_duty") }
                        return nil
                    }()
                    SlotChipRow(
                        slots: viewModel.slots,
                        selected: viewModel.selectedSlot,
                        onSelect: viewModel.selectSlot,
                        label: formatTime,
                        emptyLabel: emptyReason
                            ?? viewModel.preferredSlot.map { L("manual_slots_empty_preferred", formatTime($0)) }
                            ?? L("manual_slots_empty")
                    )
                    if viewModel.preferredUnavailable, let preferred = viewModel.preferredSlot {
                        Text(L("manual_slot_unavailable", formatTime(preferred)))
                            .font(Typo.jost(12))
                            .foregroundStyle(Color.textMuted)
                    }

                    HStack {
                        VStack(alignment: .leading, spacing: 0) {
                            Text(L("manual_sms"))
                                .font(Typo.jost(15, weight: .medium))
                                .foregroundStyle(Color.ink)
                            Text(L("manual_sms_hint"))
                                .font(Typo.jost(12))
                                .foregroundStyle(Color.textMuted)
                        }
                        Spacer()
                        BrandSwitch(isOn: $viewModel.sendSms)
                    }
                    .padding(.horizontal, 16)
                    .padding(.vertical, 12)
                    .background(RoundedRectangle(cornerRadius: 16).fill(Color.stone))

                }
                .padding(.horizontal, 20)
                .padding(.top, 12)
                .padding(.bottom, 12)
                .measureHeight($formHeight)
            }
            .scrollBounceBehavior(.basedOnSize)
            // Fuori dallo scroll: il pulsante resta sempre a vista, subito sotto il modulo.
            VStack(spacing: 8) {
                FormErrorBanner(message: viewModel.saveError)
                AccentButton(
                    text: L(viewModel.isEditing ? "manual_cta_edit" : "manual_cta"),
                    action: viewModel.save,
                    enabled: viewModel.canSave,
                    height: 56,
                    corner: 16
                )
            }
            .padding(.horizontal, 20)
            .padding(.top, 4)
            .padding(.bottom, 12)
            .measureHeight($footerHeight)
        }
        .background(Color.bone)
        .task { await viewModel.load() }
        // Alta quanto il contenuto, come "Ferie e permessi": niente buco bianco
        // fra il modulo e il pulsante. Se il modulo cresce (risultati della
        // ricerca, nuovo cliente, orari) la tendina lo segue fin sotto il notch,
        // poi scorre; trascinando non cambia altezza.
        .fittedBrandSheet(height: headerHeight + formHeight + footerHeight)
        .presentationCornerRadius(26)
        .onChange(of: viewModel.done) {
            if viewModel.done { onDismiss() }
        }
        .sheet(isPresented: $showDatePicker) {
            BrandDatePickerDialog(
                initial: viewModel.date,
                onDismiss: { showDatePicker = false },
                onConfirm: { date in
                    viewModel.selectDate(date)
                    showDatePicker = false
                }
            )
        }
    }

    private var clientList: some View {
        VStack(spacing: 0) {
            ForEach(viewModel.results.prefix(3)) { client in
                let selected = viewModel.selectedClient?.id == client.id
                Button {
                    viewModel.selectClient(client)
                } label: {
                    HStack(spacing: 11) {
                        Circle()
                            .fill(Color.bone)
                            .frame(width: 34, height: 34)
                            .overlay(
                                Text(client.initials)
                                    .font(Typo.cormorant(12, weight: .regular))
                                    .foregroundStyle(Color.ink)
                            )
                        VStack(alignment: .leading, spacing: 0) {
                            Text(client.fullName)
                                .font(Typo.titleSmall)
                                .foregroundStyle(Color.ink)
                                .lineLimit(1)
                            Text("\(client.phone) · \(L("manual_visits", client.visitCount))")
                                .font(Typo.jost(12))
                                .foregroundStyle(Color.textMuted)
                                .lineLimit(1)
                        }
                        Spacer()
                    }
                    .padding(.horizontal, 14)
                    .padding(.vertical, 12)
                    .background(selected ? Color.oliveWood.opacity(0.18) : Color.stone)
                }
                .buttonStyle(.plain)
            }
            Button(action: viewModel.startCreateClient) {
                HStack(spacing: 11) {
                    Circle()
                        .fill(Color.bone)
                        .frame(width: 34, height: 34)
                        .overlay(
                            Image(systemName: "plus")
                                .font(.system(size: 13))
                                .foregroundStyle(Color.oliveWood)
                        )
                    Text(L("manual_new_client"))
                        .font(Typo.bodyMedium)
                        .foregroundStyle(Color.textMuted)
                    Spacer()
                }
                .padding(.horizontal, 14)
                .padding(.vertical, 12)
            }
            .buttonStyle(.plain)
        }
        .background(RoundedRectangle(cornerRadius: 16).fill(Color.stone))
        .clipShape(RoundedRectangle(cornerRadius: 16))
    }


    /// Stone field that reads as a picker: value plus a chevron, as in the mockup.
    private func dropdownField(_ value: String, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            dropdownLabel(value)
        }
        .buttonStyle(.plain)
    }

    private func dropdownLabel(_ value: String) -> some View {
        HStack {
            Text(value)
                .font(Typo.bodyLarge)
                .foregroundStyle(Color.ink)
                .lineLimit(1)
            Spacer()
            Image(systemName: "chevron.down")
                .font(.system(size: 13))
                .foregroundStyle(Color.textMuted)
        }
        .padding(.horizontal, 14)
        .frame(height: 56)
        .background(RoundedRectangle(cornerRadius: 14).fill(Color.stone))
        .contentShape(Rectangle())
    }
}

/// "Luca Ferrante" -> "Luca F."
func shortOperatorName(_ name: String) -> String {
    let parts = name.split(separator: " ").map(String.init)
    guard parts.count > 1, let initial = parts[1].first else { return name }
    return "\(parts[0]) \(initial)."
}

File: android/feature/admin/src/main/kotlin/com/devora/mencare/feature/admin/manual/ManualBookingScreen.kt
package com.devora.mencare.feature.admin.manual

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Add
import androidx.compose.material.icons.outlined.ArrowDropDown
import androidx.compose.material.icons.outlined.Check
import androidx.compose.material.icons.outlined.Search
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.devora.mencare.core.common.formatDateShort
import com.devora.mencare.core.common.formatDuration
import com.devora.mencare.core.common.formatTime
import com.devora.mencare.core.designsystem.component.AccentButton
import com.devora.mencare.core.designsystem.component.BarAction
import com.devora.mencare.core.designsystem.component.BrandDatePickerDialog
import com.devora.mencare.core.designsystem.component.BrandSectionLabel
import com.devora.mencare.core.designsystem.component.BrandSwitch
import com.devora.mencare.core.designsystem.component.DropdownOption
import com.devora.mencare.core.designsystem.component.FilledTextField
import com.devora.mencare.core.designsystem.component.MultiSelectDropdown
import com.devora.mencare.core.designsystem.component.NameField
import com.devora.mencare.core.designsystem.component.PhoneField
import com.devora.mencare.core.designsystem.component.SlotChipRow
import com.devora.mencare.core.designsystem.theme.Bone
import com.devora.mencare.core.designsystem.theme.Cormorant
import com.devora.mencare.core.designsystem.theme.ErrorRed
import com.devora.mencare.core.designsystem.theme.Ink
import com.devora.mencare.core.designsystem.theme.OliveLight
import com.devora.mencare.core.designsystem.theme.OliveTint
import com.devora.mencare.core.designsystem.theme.OliveWood
import com.devora.mencare.core.designsystem.theme.Radii
import com.devora.mencare.core.designsystem.theme.Stone
import com.devora.mencare.core.designsystem.theme.StoneBorder
import com.devora.mencare.core.designsystem.theme.TextMuted
import com.devora.mencare.core.ui.crm.FixedClientRow
import com.devora.mencare.feature.admin.R
import java.time.LocalDate
import java.time.LocalTime

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ManualBookingScreen(
    onDismiss: () -> Unit,
    operatorId: String? = null,
    date: LocalDate = LocalDate.now(),
    time: LocalTime? = null,
    // "Modifica" dall'agenda: modulo precompilato con questo appuntamento.
    editAppointment: com.devora.mencare.core.model.Appointment? = null,
    editClient: com.devora.mencare.core.model.ClientRecord? = null,
    // Una chiave per apertura: ogni sheet riparte da un modulo vuoto.
    sessionKey: String = "manual-booking",
    viewModel: ManualBookingViewModel = hiltViewModel(key = sessionKey),
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    var showDatePicker by rememberSaveable { mutableStateOf(false) }
    var operatorMenuOpen by remember { mutableStateOf(false) }

    LaunchedEffect(viewModel) {
        if (editAppointment != null) {
            viewModel.applyEdit(editAppointment, editClient)
        } else {
            viewModel.applyPreset(operatorId, date, time)
        }
    }

    LaunchedEffect(state.done) {
        if (state.done) onDismiss()
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = Bone,
        shape = RoundedCornerShape(topStart = 26.dp, topEnd = 26.dp),
    ) {
        Column(Modifier.navigationBarsPadding()) {
            Row(
                modifier = Modifier.fillMaxWidth().padding(horizontal = 20.dp, vertical = 6.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Box(Modifier.width(72.dp), contentAlignment = Alignment.CenterStart) {
                    BarAction(stringResource(R.string.manual_cancel), onDismiss, color = Ink)
                }
                Text(
                    stringResource(
                        if (state.editingId != null) R.string.manual_title_edit else R.string.manual_title_short,
                    ),
                    fontFamily = Cormorant,
                    fontSize = 21.sp,
                    color = Ink,
                    textAlign = TextAlign.Center,
                    maxLines = 1,
                    modifier = Modifier.weight(1f),
                )
                Box(Modifier.width(72.dp), contentAlignment = Alignment.CenterEnd) {
                    BarAction(stringResource(R.string.manual_save), viewModel::save, color = OliveLight)
                }
            }

            Column(
                modifier = Modifier
                    .weight(1f, fill = false)
                    .verticalScroll(rememberScrollState())
                    .imePadding()
                    .padding(horizontal = 20.dp)
                    .padding(top = 12.dp, bottom = 12.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                BrandSectionLabel(stringResource(R.string.manual_client))
                // In modifica il cliente è quello dell'appuntamento: si mostra e basta.
                if (state.editingId != null) {
                    FixedClientRow(state.selectedClient)
                } else {
                FilledTextField(
                    value = state.query,
                    onValueChange = viewModel::search,
                    label = "",
                    placeholder = stringResource(R.string.manual_search_hint),
                    leadingIcon = Icons.Outlined.Search,
                )
                Column(
                    modifier = Modifier.fillMaxWidth().clip(Radii.Md).background(Bone).border(1.5.dp, StoneBorder, Radii.Md),
                ) {
                    state.results.take(3).forEach { client ->
                        val selected = state.selectedClient?.id == client.id
                        Row(
                            modifier = Modifier
                                .fillMaxWidth()
                                .background(if (selected) OliveTint else Bone)
                                .clickable { viewModel.selectClient(client) }
                                .padding(horizontal = 14.dp, vertical = 12.dp),
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            Box(
                                modifier = Modifier.size(34.dp).clip(CircleShape).background(Bone),
                                contentAlignment = Alignment.Center,
                            ) {
                                Text(
                                    "${client.firstName.first()}${client.lastName.first()}",
                                    fontFamily = Cormorant,
                                    fontSize = 12.sp,
                                    color = Ink,
                                )
                            }
                            Column(Modifier.weight(1f).padding(start = 11.dp)) {
                                Text(
                                    "${client.firstName} ${client.lastName}",
                                    style = MaterialTheme.typography.titleSmall,
                                    color = Ink,
                                    maxLines = 1,
                                    overflow = TextOverflow.Ellipsis,
                                )
                                Text(
                                    "${client.phone} · ${stringResource(R.string.manual_visits, client.visitCount)}",
                                    style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.sp),
                                    color = TextMuted,
                                    maxLines = 1,
                                    overflow = TextOverflow.Ellipsis,
                                )
                            }
                        }
                    }
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .clickable(onClick = viewModel::startCreateClient)
                            .padding(horizontal = 14.dp, vertical = 12.dp),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Box(
                            modifier = Modifier.size(34.dp).clip(CircleShape).background(Bone),
                            contentAlignment = Alignment.Center,
                        ) {
                            Icon(
                                Icons.Outlined.Add,
                                contentDescription = null,
                                tint = OliveWood,
                                modifier = Modifier.size(16.dp),
                            )
                        }
                        Text(
                            stringResource(R.string.manual_new_client),
                            style = MaterialTheme.typography.bodyMedium,
                            color = TextMuted,
                            modifier = Modifier.padding(start = 11.dp),
                        )
                    }
                }
                }

                if (state.creatingClient) {
                    Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                        NameField(
                            value = state.newFirst,
                            onValueChange = { viewModel.setNewClientField("first", it) },
                            label = stringResource(R.string.manual_new_client_first),
                            error = state.newFirstError,
                            modifier = Modifier.weight(1f),
                        )
                        NameField(
                            value = state.newLast,
                            onValueChange = { viewModel.setNewClientField("last", it) },
                            label = stringResource(R.string.manual_new_client_last),
                            error = state.newLastError,
                            modifier = Modifier.weight(1f),
                        )
                    }
                    PhoneField(
                        value = state.newPhone,
                        onValueChange = { viewModel.setNewClientField("phone", it) },
                        label = stringResource(R.string.manual_new_client_phone),
                        error = state.newPhoneError,
                    )
                    if (state.newClientFailed) {
                        Text(
                            stringResource(R.string.manual_error_new_client),
                            style = MaterialTheme.typography.bodySmall,
                            color = ErrorRed,
                        )
                    }
                    AccentButton(
                        text = stringResource(R.string.manual_new_client),
                        onClick = viewModel::createClient,
                        loading = state.creatingClientBusy,
                        height = 48.dp,
                        shape = RoundedCornerShape(16.dp),
                    )
                }

                Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                    Column(Modifier.weight(1f)) {
                        BrandSectionLabel(stringResource(R.string.manual_operator))
                        Spacer(Modifier.height(8.dp))
                        val current = state.operators.firstOrNull { it.id == state.selectedOperatorId }
                        // Menu a tendina: tutti gli operatori, spunta su quello scelto.
                        Box {
                            DropdownField(
                                value = current?.let { shortOperatorName(it.name) } ?: "—",
                            ) { operatorMenuOpen = true }
                            DropdownMenu(
                                expanded = operatorMenuOpen,
                                onDismissRequest = { operatorMenuOpen = false },
                                containerColor = Bone,
                            ) {
                                state.operators.forEach { operator ->
                                    DropdownMenuItem(
                                        text = { Text(operator.name, color = Ink) },
                                        onClick = {
                                            operatorMenuOpen = false
                                            viewModel.selectOperator(operator.id)
                                        },
                                        trailingIcon = {
                                            if (operator.id == state.selectedOperatorId) {
                                                Icon(Icons.Outlined.Check, contentDescription = null, tint = OliveWood)
                                            }
                                        },
                                    )
                                }
                            }
                        }
                    }
                    Column(Modifier.weight(1f)) {
                        BrandSectionLabel(stringResource(R.string.manual_date))
                        Spacer(Modifier.height(8.dp))
                        DropdownField(value = formatDateShort(state.date)) {
                            showDatePicker = true
                        }
                    }
                }

                BrandSectionLabel(stringResource(R.string.manual_services))
                // Tendina multi-selezione, gemella di quella dell'operatore.
                MultiSelectDropdown(
                    options = state.services.map { DropdownOption(it.id, it.name) },
                    selectedIds = state.selectedServiceIds,
                    onToggle = viewModel::toggleService,
                    placeholder = stringResource(R.string.manual_services_placeholder),
                )

                BrandSectionLabel(
                    stringResource(R.string.manual_slots, formatDuration(state.totalMinutes)),
                )
                // Colonna vuota non vuol dire libera: se l'operatore non è in turno
                // o non esegue i servizi scelti, lo si dice.
                val emptyReason = when {
                    state.selectedOperatorId == null || state.selectedServiceIds.isEmpty() -> null
                    state.operatorIneligible -> stringResource(R.string.manual_slots_ineligible)
                    state.operatorOffDuty -> stringResource(R.string.manual_slots_off_duty)
                    else -> null
                }
                SlotChipRow(
                    slots = state.slots,
                    selected = state.selectedSlot,
                    onSelect = viewModel::selectSlot,
                    label = ::formatTime,
                    emptyLabel = emptyReason
                        ?: state.preferredSlot
                            ?.let { stringResource(R.string.manual_slots_empty_preferred, formatTime(it)) }
                        ?: stringResource(R.string.manual_slots_empty),
                )
                val preferred = state.preferredSlot
                if (state.preferredUnavailable && preferred != null) {
                    Text(
                        stringResource(R.string.manual_slot_unavailable, formatTime(preferred)),
                        style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.sp),
                        color = TextMuted,
                    )
                }

                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clip(Radii.Md)
                        .background(Bone)
                        .border(1.5.dp, StoneBorder, Radii.Md)
                        .padding(horizontal = 16.dp, vertical = 12.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Column(Modifier.weight(1f)) {
                        Text(
                            stringResource(R.string.manual_sms),
                            style = MaterialTheme.typography.titleMedium.copy(fontSize = 15.sp),
                            color = Ink,
                        )
                        Text(
                            stringResource(R.string.manual_sms_hint),
                            style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.sp),
                            color = TextMuted,
                        )
                    }
                    BrandSwitch(checked = state.sendSms, onCheckedChange = viewModel::setSendSms)
                }

                state.bookingError?.let { error ->
                    Text(
                        stringResource(
                            when (error) {
                                ManualBookingError.SLOT_TAKEN -> R.string.manual_error_slot_taken
                                ManualBookingError.GENERIC -> R.string.manual_error_save
                            },
                        ),
                        style = MaterialTheme.typography.bodySmall,
                        color = ErrorRed,
                    )
                }

                AccentButton(
                    text = stringResource(
                        if (state.editingId != null) R.string.manual_cta_edit else R.string.manual_cta,
                    ),
                    onClick = viewModel::save,
                    enabled = state.canSave,
                    loading = state.saving,
                    height = 56.dp,
                    shape = RoundedCornerShape(16.dp),
                )
            }
        }
    }

    if (showDatePicker) {
        BrandDatePickerDialog(
            initial = state.date,
            onDismiss = { showDatePicker = false },
            onConfirm = {
                viewModel.selectDate(it)
                showDatePicker = false
            },
        )
    }
}

/** Stone field that reads as a picker: value plus a chevron, as in the mockup. */
@Composable
private fun DropdownField(value: String, onClick: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(56.dp)
            .clip(Radii.Md)
            .background(Bone)
            .border(1.5.dp, StoneBorder, Radii.Md)
            .clickable(onClick = onClick)
            .padding(horizontal = 14.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            value,
            style = MaterialTheme.typography.bodyLarge,
            color = Ink,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.weight(1f),
        )
        Icon(
            Icons.Outlined.ArrowDropDown,
            contentDescription = null,
            tint = TextMuted,
            modifier = Modifier.size(20.dp),
        )
    }
}

/** "Luca Ferrante" -> "Luca F." */
private fun shortOperatorName(name: String): String {
    val parts = name.split(' ').filter { it.isNotBlank() }
    return if (parts.size > 1) "${parts.first()} ${parts[1].first()}." else name
}
